app_dart/lib/src/request_handling/subscription_handler.dart

```diff
@@ -19,23 +25,124 @@ import 'body.dart';
+abstract class SubscriptionHandler extends RequestHandler<Body> {
   /// Creates a new [SubscriptionHandler].
   const SubscriptionHandler({
+    required this.cache,
     required Config config,
-    required AuthenticationProvider authenticationProvider,
-  }) : super(config: config, authenticationProvider: authenticationProvider);
+    required this.authProvider,
+    required this.topicName,
+  }) : super(
+          config: config,
+        );
 
-  Future<String> get _requestString async {
-    final String requestString = String.fromCharCodes(requestBody!);
-    log.fine(requestString);
-    return requestString;
+  final CacheService cache;
+
+  /// Service responsible for authenticating this [HttpRequest].
+  final AuthenticationProvider authProvider;
+
+  /// Unique identifier of the PubSub in this cloud project.
+  final String topicName;
+
+  /// The authentication context associated with the HTTP request.
+  ///
+  /// This is guaranteed to be non-null. If the request was unauthenticated,
+  /// the request will be denied.
+  @protected
+  AuthenticatedContext get authContext => getValue<AuthenticatedContext>(ApiKey.authContext)!;
+
+  /// The [PushMessage] from this [HttpRequest].
+  @protected
+  PushMessage get message => getValue<PushMessage>(PubSubKey.message)!;
+
+  @override
+  Future<void> service(HttpRequest request) async {
+    AuthenticatedContext authContext;
+    try {
+      authContext = await authProvider.authenticate(request);
+    } on Unauthenticated catch (error) {
+      final HttpResponse response = request.response;
+      response
+        ..statusCode = HttpStatus.unauthorized
+        ..write(error.message);
+      await response.flush();
+      await response.close();
+      return;
+    }
+
+    List<int> body;
+    try {
+      body = await request.expand<int>((List<int> chunk) => chunk).toList();
+    } catch (error) {
+      final HttpResponse response = request.response;
+      response
+        ..statusCode = HttpStatus.internalServerError
+        ..write('$error');
+      await response.flush();
+      await response.close();
+      return;
+    }
+
+    PushMessageEnvelope? envelope;
+    if (body.isNotEmpty) {
+      try {
+        final Map<String, dynamic> json = jsonDecode(utf8.decode(body)) as Map<String, dynamic>;
+        envelope = PushMessageEnvelope.fromJson(json);
+      } catch (error) {
+        final HttpResponse response = request.response;
+        response
+          ..statusCode = HttpStatus.internalServerError
+          ..write('$error');
+        await response.flush();
+        await response.close();
+        return;
+      }
+    }
+
+    if (envelope == null) {
+      throw const BadRequestException('Failed to get message');
+    }
+    log.finer(envelope.toJson());
+
+    final String messageId = envelope.message!.messageId!;
+
+    final Uint8List? messageLock = await cache.getOrCreate(topicName, messageId);
```

Comment:
Do we need an implementation that allows to specify a TTL? ideally the TTL should match the configuration in the pub/sub subscription

Reply:
No, I'm referring for the TTL for the item in the cache. One week seems like too for messages that will expire in 24h with the default pubsub configurations. It would be great if we can add those optimizations now that we have the context fresh.

Reply:
I'll decrease this to 1 day, but a message id is guaranteed to be unique over the lifetime of a topic. By default that allows for 64^19 messages, and PubSub will dynamically increase the range to guarantee this.